File: frontend/types/profiling.ts
// frontend/types/profiling.ts
// TypeScript types for data profiling and quality reporting.

export interface ColumnProfile {
  name: string;
  data_type: string;
  total_count: number;
  null_count: number;
  null_percentage: number;
  unique_count: number;
  distinct_percentage: number;
  min_value?: string | number | null;
  max_value?: string | number | null;
  mean_value?: number | null;
  std_dev?: number | null;
  quantiles?: { "25%": number; "50%": number; "75%": number } | null;
  top_frequencies?: { value: string; count: number }[] | null;
  outlier_count?: number;
}

export interface DataQualityReport {
  dataset_id: string;
  dataset_name: string;
  total_rows: number;
  total_columns: number;
  health_score: number;
  overall_null_percentage: number;
  duplicate_row_count: number;
  column_profiles: ColumnProfile[];
  warnings: string[];
}

File: backend/app/schemas/profiling.py
"""
backend/app/schemas/profiling.py
Pydantic schemas for automated data profiling and quality reporting.
"""
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field


class ColumnProfile(BaseModel):
    name: str
    data_type: str
    total_count: int
    null_count: int
    null_percentage: float
    unique_count: int
    distinct_percentage: float
    min_value: Optional[Any] = None
    max_value: Optional[Any] = None
    mean_value: Optional[float] = None
    std_dev: Optional[float] = None
    quantiles: Optional[Dict[str, float]] = None  # e.g. {"25%": x, "50%": y, "75%": z}
    top_frequencies: Optional[List[Dict[str, Any]]] = None
    outlier_count: Optional[int] = 0


class DataQualityReport(BaseModel):
    dataset_id: str
    dataset_name: str
    total_rows: int
    total_columns: int
    health_score: float  # 0 to 100%
    overall_null_percentage: float
    duplicate_row_count: int
    column_profiles: List[ColumnProfile]
    warnings: List[str]
